File: C++/Design/1146. Snapshot Array.cpp
#include <bits/stdc++.h>


using namespace std;

class SnapshotArray {
private:
    vector<unordered_map<int, int>> matrix;
    int curSnapId = 0;
public:
    SnapshotArray(int length) {
        matrix.resize(length);
        for (int i = 0; i < length; ++i)
            matrix[i][curSnapId] = 0;
    }

    void set(int index, int val) {
        matrix[index][curSnapId] = val;
    }

    int snap() {
        return curSnapId++;
    }

    int get(int index, int snap_id) {
        while (matrix[index].find(snap_id) == matrix[index].end())
            snap_id--;
        return matrix[index][snap_id];
    }
};

File: Python/hash_map/1146. Snapshot Array.py
class SnapshotArray:

    def __init__(self, length: int):
        self.cur_snap_id = 0
        self.matrix = [{} for _ in range(length)]

        for i in range(length):
            self.matrix[i][0] = 0

    def set(self, index: int, val: int) -> None:
        self.matrix[index][self.cur_snap_id] = val

    def snap(self) -> int:
        self.cur_snap_id += 1
        return self.cur_snap_id - 1

    def get(self, index: int, snap_id: int) -> int:
        while snap_id not in self.matrix[index]:
            snap_id -= 1
        return self.matrix[index][snap_id]
